### PR TITLE
API:Add the usage in javadoc of ValidationException(#3895)

### PR DESCRIPTION
**Pull request by dylzlt**
I add the usage in javadoc of ValidationException.
Can you give me some suggestion if my description is not accurate.
thanks! @jackye1995

### CODE REVIEW

**Comment by szehon-ho** [comment]
Not sure if we should put developer-facing comments here, as this is a user-facing Javadoc..   Maybe a better place is the Contribution guide?  https://iceberg.apache.org/contribute/

**Comment by szehon-ho** [comment]
Sorry I just read the original issue as it wasnt linked (#3895 ) , I get it now and think would be great to have a Javadoc on this class.

I think the problem is that this proposed Javadoc is developer-facing, and should be user-facing.  Specifically, the 'we' here seems like its meant for developer to know what exception to throw... but it should be instead describing to user what situations to expect it from.

**Comment by dylzlt** [comment]
> Sorry I just read the original issue as it wasnt linked (#3895 ) , I get it now and think would be great to have a Javadoc on this class.
> 
> I think the problem is that this proposed Javadoc is developer-facing, and should be user-facing. Specifically, the 'we' here seems like its meant for developer to know what exception to throw... but it should be instead describing to user what situations to expect it from.

Oh, thanks for your suggestion. You mean the Javadoc is used to tell user  arguments inconsistencies when they catch the ValidationException, rather than tell developer how to use it(because this is developer rule)?Is my understanding right?

**Comment by szehon-ho** [comment]
Yea, that's what I am thinking, because this will be on Iceberg's user-facing Javadocs.  Also in general we try to keep documentation free from personal pronouns (like we, I..) and just state what the class does.

**Comment by dylzlt** [comment]
> Yea, that's what I am thinking, because this will be on Iceberg's user-facing Javadocs. Also in general we try to keep documentation free from personal pronouns (like we, I..) and just state what the class does.

Thank you very much for your explanation, I modify the doc, please take another look for me，thanks！

**Comment by szehon-ho** [comment]
I think it needs a bit more context, how about something like:


"Exception raised when validation checks fail.  A ValidationException occurs from inconsistent state that would cause an operation to abort, as opposed to an {@link IllegalArgumentException}, which occurs from invalid arguments."


I would also love to add another example here:



"For example, this is thrown when attempting to create a table with a {@link PartitionSpec} that is not compatible with the table {@link Schema}, or attempting to add or remove files to a table that would violate the configured isolation level due to concurrent commits."



That being said, I did not raise the original issue , so maybe @rdblue @jackye1995 have more thoughts, or @aokolnychyi @RussellSpitzer @kbendick  as well :)

**Comment by dylzlt** [comment]
> 

OK, I agree that the description is better! 
But there is a problem that the IsolationLevel is in iceberg-core module rather than api(the link is invalid), so remove the link will be better?
the example is here:
“For example, this is thrown when attempting to create a table with a {@link PartitionSpec} that is not compatible with the table {@link Schema}, or attempting to add or remove files to a table that would violate the configured IsolationLevel due to concurrent commits.”

**Comment by szehon-ho** [comment]
> But there is a problem that the IsolationLevel is in iceberg-core module rather than api(the link is invalid)
Yea I think you can qualify the class?  Or else removing it is also fine with me, as it's not a big deal.

File: api/src/main/java/org/apache/iceberg/exceptions/ValidationException.java [hunk 1]
```diff
@@ -24,7 +24,8 @@
 import org.apache.iceberg.Schema;
 
 /**
- * Exception raised when validation checks fail.
+ * Exception raised when arguments inconsistencies checks fail.
```
[the remark this answers is not in the document]
Comment (dylzlt) [reply api/src/main/java/org/apache/iceberg/exceptions/ValidationException.java]: Thank you for that. I have modified that.

File: api/src/main/java/org/apache/iceberg/exceptions/ValidationException.java [hunk 2]
```diff
@@ -24,7 +24,8 @@
...
+ * {@link IllegalArgumentException} raised when arguments aren't allowed.
```
[the remark this answers is not in the document]
Comment (dylzlt) [reply api/src/main/java/org/apache/iceberg/exceptions/ValidationException.java]: ok, I have modified that.Thank you for your review

**Comment by szehon-ho** [comment]
@dylzlt sorry for the delay, I synced with @RussellSpitzer to reduce the back and forth, his main concern was to make sure we say IllegalArgumentException means it is never valid as opposed to it's not valid now because of other arguments.

So can you do something like?
```
Exception which is raised when the arguments are valid in isolation, but not in conjunction with other arguments or state, as opposed to {@link IllegalArgumentException} which is raised when an argument value is always invalid.  A ValidationException will cause the operation to abort.
```

I think it should be good to go after this.

**Comment by dylzlt** [comment]
> 

ok, I get it. Thank you for your patience

**Comment by dylzlt** [comment]
> @dylzlt sorry for the delay, I synced with @RussellSpitzer to reduce the back and forth, his main concern was to make sure we say IllegalArgumentException means it is never valid as opposed to it's not valid now because of other arguments.
> 
> So can you do something like?
> 
> ```
> Exception which is raised when the arguments are valid in isolation, but not in conjunction with other arguments or state, as opposed to {@link IllegalArgumentException} which is raised when an argument value is always invalid.  A ValidationException will cause the operation to abort.
> ```
> 
> I think it should be good to go after this.
ok, I get it. Thank you for your patience. I complete it.

File: api/src/main/java/org/apache/iceberg/exceptions/ValidationException.java [hunk 3]
```diff
@@ -24,7 +24,11 @@
 import org.apache.iceberg.Schema;
 
 /**
- * Exception raised when validation checks fail.
+ * Exception which is raised when the arguments are valid in isolation,
```
Comment (szehon-ho) [inline api/src/main/java/org/apache/iceberg/exceptions/ValidationException.java:27]: Small nit, can you try to make each line be longer and try to line up towards the right?  

I don't see many places where the doc line-breaks on a comma, usually it's on a period.  You can also add a <p> after the period.

A good example in my opinion : https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/ExpireSnapshots.java

Thank you

Comment (dylzlt) [reply api/src/main/java/org/apache/iceberg/exceptions/ValidationException.java]: ok, I‘m sorry for it.
I just want to make a part of sentence is complete in one line.

**Review by szehon-ho — APPROVED** [review]

**Comment by szehon-ho** [comment]
I accidentally mis-labelled it as Core from Api when simplifying the commit message, sorry about this

**Pull request merged**